[Gabriel4357] iP

Gabriel4357 opened the pull request:
# DukePro 
> “The wages of sin is **debugging** ” - Ron Jeffries.  

Feature of DukePro include 
* adding tasks to your list 
* set deadlines to the tasks 
* set events when it begins and when it ends 
* track these tasks and mark them when done 

Reasons to use DukePro 
1. Easy to use 
2. Easy to pick up
3. Free of charge 
4. Very useful in managing tasks 

Take a look at the TaskList code : 

```java
public class TaskList extends ArrayList<Task> {
}
``` 

Example of how Duke works : 
- [ ] CS2101 Slides 
- [ ] CS2100 Tutorial 
- [X] CS2103T iP 
- [ ] ST2334 Tutorial 
- [ ] \(Optional) *CFG1003 Slides*

👍 😄 😃 🏆 

You can also use [Google](https://www.google.co.uk/) to search for the **best** todo apps and DukePro will show up as 
🥇 . 

Or you could visit [my Github repo](https://github.com/Gabriel4357/ip) to find out more!

Currently, we have `class Todo ` , `class Deadline` and `class Event`. 

~~However, it is currently buggy and not ready to be mass produced.~~

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Gabriel4357 committed:
changed name, and removed tag

Gabriel4357 committed:
implement initial skeletal version

Gabriel4357 committed:
improved skeletal version

Gabriel4357 committed:
customisation added

Gabriel4357 committed:
use of formatted string

Gabriel4357 committed:
added an array and counter variable

Gabriel4357 committed:
added useful methods

Gabriel4357 committed:
logic added

Gabriel4357 committed:
added task class

Gabriel4357 committed:
duke package created

Gabriel4357 committed:
task class updated

Gabriel4357 committed:
updated logic in duke

Gabriel4357 committed:
logic of mark and unmark added

Gabriel4357 committed:
new types of tasks added

Gabriel4357 committed:
logic of duke revamped

Gabriel4357 committed:
todo logic ready

Gabriel4357 committed:
deadline logic ready

Gabriel4357 committed:
event logic working

Gabriel4357 committed:
tests set up

Gabriel4357 committed:
UI Test done

Gabriel4357 committed:
clean up logic

Gabriel4357 committed:
use arraylist now, and finish delete

Gabriel4357 committed:
both errors added

Gabriel4357 committed:
delete added

Gabriel4357 committed:
bugs fixed

Gabriel4357 committed:
final changes

Gabriel4357 committed:
Refactor the method name to printline()

Gabriel4357 committed:
Use a file to track past data.

Cannot save a list of items.

Can store past data is needed as a to-do application.

Let's,
*create a file that commands will be written to
* read from that file and calls the commands at the start of each run

Gabriel4357 committed:
Align code to match coding style

Gabriel4357 committed:
Fix merge conflict

Gabriel4357 committed:
Introduce LocalTimeDate into the application

Event and Deadline class uses Strings as their parameters.

Using actual times will be more accurate to a todo application.

Let's
*change the strings into LocalTimeDates

Gabriel4357 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Introduce LocalTimeDate into the application

Gabriel4357 committed:
Remove counter

Gabriel4357 committed:
Change code to include even more OOP than before.

Let's
*add ui, storage, parser, and tasklist as a means of abstraction

Gabriel4357 committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Change code to include even more OOP than before.
  Remove counter

Gabriel4357 committed:
Touch up code

Gabriel4357 committed:
Add package 'duke'

Gabriel4357 committed:
Merge branch 'branch-A-Packages'

* branch-A-Packages:
  Add package 'duke'

Gabriel4357 committed:
Merge commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552'

* commit 'ed6d4d25bd4e34461919ffe785a6f10f4c087552':
  Bump gradle and lib version
  Add Gradle support

Gabriel4357 committed:
Add Gradle to the working project.

Gabriel4357 committed:
Merge branch 'branch-A-Gradle-'

* branch-A-Gradle-:
  Add Gradle to the working project.

Gabriel4357 committed:
Add more OOP into code.

UI does not handle all interactions with user.

Abstracted out Parser sending messages to user and leaving it to
UI to send the responses.

Let's,
*add a method sendMessage() in UI to handle the sending
of messages to user.

Gabriel4357 committed:
Merge branch 'branch-A-MoreOOP'

* branch-A-MoreOOP:
  Add more OOP into code.

Gabriel4357 committed:
Fix merge conflict

Gabriel4357 committed:
Add JUnit Tests

No testings done to test the methods and classes

Having automated test cases makes it easier to spot
bugs when the code changes.

Let's,
*add JUnit tests: ParserTest,
TaskTest and TaskListTest

Gabriel4357 committed:
Merge branch 'branch-A-JUnit'

* branch-A-JUnit:
  Add JUnit Tests
  Fix merge conflict

Gabriel4357 committed:
Package the app as a Jar file

Gabriel4357 committed:
Add JavaDocs

Gabriel4357 committed:
Fix code to coding standards

Gabriel4357 committed:
Add a find command

Not possible to find a few tasks that contains
a particular string

Find command makes it easier
to find particular tasks

Gabriel4357 committed:
Merge branch 'branch-A-Jar'

* branch-A-Jar:
  Package the app as a Jar file

Gabriel4357 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDocs

Gabriel4357 committed:
Merge conflict fixed

Gabriel4357 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add a find command

Gabriel4357 committed:
Fix bug on not quitting on 'bye' command

Gabriel4357 committed:
Change parse to return a String instead

itsNatTan commented on src/main/java/duke/Parser.java:
Could the actual code be abstracted out into a different class?
Perhaps using switch would be better?

WZWren replied:
I agree with this - since there is so many cases, `switch` would make this easier to work with.

itsNatTan commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,75 @@
+        fw.write(textToAppend);
+        fw.close();
+    }
+    public static void preloadFromFile(TaskList list) {
```
Very good idea to make a preloadFromFile function! I did not think of that. Yoinked

itsNatTan commented on src/main/java/duke/Tag.java:
Could this be placed in the Task class? Or is this used in other places as well

itsNatTan reviewed (commented):
Very well done! I think the code is clean and mostly easily understandable. Would prefer switch instead of if-else for readability. Good initiative on some of the functions. 10/10

Gabriel4357 committed:
Implement a working GUI

chuababyy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,210 @@
+                int numToUnmark = Integer.parseInt(splitText[1]) - 1;
+                list.get(numToUnmark).markAsIncomplete();
+
+                ui.printline();
+                ui.sendMessage("Alright! I'll uncheck this task for you: ");
+                ui.sendMessage(String.format("\t [%s] [%s] %s", list.get(numToUnmark).tag,
+                        list.get(numToUnmark).getStatusIcon(), list.get(numToUnmark)));
+                ui.printline();
```
You might want to consider extracting all these out and putting it in a function in the Ui class. This can help your Parser class to be easier to understand. Same comment for each of the conditions below. 

For example, in your Ui class you can have a show `UnmarkMessage()` method etc.

chuababyy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,210 @@
+                } else {
+                    boolean hasSearch = false;
+                    boolean isFirst = true;
+                    for (int i = 0; i < list.size(); i++) {
+                        if (list.get(i).toString().contains(search)) {
+                            hasSearch = true;
+                            if (hasSearch == true && isFirst == true) {
+                                ui.printline();
+                                ui.sendMessage("Here are the matching tasks in your list:");
+                                ui.sendMessage(String.format("\t [%s] [%s] %s", list.get(i).tag,
+                                        list.get(i).getStatusIcon(), list.get(i).toString()));
+                                isFirst = false;
+                            } else if (hasSearch == true && isFirst == false) {
+                                ui.sendMessage(String.format("\t [%s] [%s] %s", list.get(i).tag,
+                                        list.get(i).getStatusIcon(), list.get(i).toString()));
+                            }
+                        }
+                    }
```
Can this function of finding a task be handled in the TaskList class instead?

chuababyy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,210 @@
+                        ui.printline();
+                        ui.sendMessage("Noted Sir. I've added this task to your list: ");
+                        ui.sendMessage(String.format("\t [%s] [%s] %s", event.tag,
+                                event.getStatusIcon(),
+                                event.toString()));
```
Can `getStatusIcon()` be included as a part of the `toString()` method? ie the `toString()` method for `ToDo` can return [X] description. Instead of just returning the description.

chuababyy commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,13 @@
+package duke;
```
Perhaps you might want to consider using this class help to handle more situations so that the Parser class does not have to handle each case individually

Gabriel4357 committed:
Refine GUI and add more errors thrown

Gabriel4357 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Refine GUI and add more errors thrown
  Implement a working GUI
  Change parse to return a String instead

juliusgambe commented on src/main/java/duke/DialogBox.java:
```diff
@@ -0,0 +1,61 @@
+ * This control represents a dialog box consisting of an ImageView to represent the speaker's face and a label
+ * containing text from the speaker.
+ */
+public class DialogBox extends HBox {
+    @FXML
+    private Label dialog;
+    @FXML
+    private ImageView displayPicture;
+
+    private DialogBox(String text, Image img) {
+        try {
+            FXMLLoader fxmlLoader = new FXMLLoader(MainWindow.class.getResource("/view/DialogBox.fxml"));
+            fxmlLoader.setController(this);
+            fxmlLoader.setRoot(this);
+            fxmlLoader.load();
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        dialog.setText(text);
+        displayPicture.setImage(img);
+    }
+
```
hello can make ur code better?!?!?!?!

WZWren commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+ */
+public class Deadline extends Task{
+
+    protected LocalDateTime deadline;
```
Is there a reason for this to be `protected` rather than `private`? I noticed this pattern in a few other classes too.

WZWren commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,50 @@
+                deadline.getYear() + ", " + deadline.toLocalTime();
+    }
+
+    /**
+     * @inheritDoc
+     */
```
Nice use of the `inheritDoc` tag!

WZWren commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,58 @@
+     * @inheritDoc
+     */
+    @Override
+    public String toString() {
+        return description;
+    }
```
I think the `toString` method for Task is underutilized. Maybe consider using it to also display the `getStatusIcon`?

WZWren commented on src/main/java/duke/UI.java:
```diff
@@ -0,0 +1,75 @@
+ *
+ * It is in charge responding to the user.
+ */
+public class UI {
```
Perhaps you should use the name Ui instead of UI, to match the tendency to associate abbreviated names as 1 word?

WZWren reviewed (commented):
LGTM, just some minor nits regarding naming and accessor properties.

_Note: this isn't related to the coding standard, but perhaps you should .gitignore your class binaries? Clones can generate class files from your source code directly._

hyc17003 commented on src/test/java/duke/TaskListTest.java:
```diff
@@ -0,0 +1,37 @@
+        assertEquals(taskListTest.get(2), event);
+    }
+    @Test
+    public void deleteTest() {
```
Maybe naming of this test function could be clearer? There might be a bit of ambiguity with the current name, maybe it can be something like 'deleteFunctionTest'.

hyc17003 commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,53 @@
+    /**
+     * Changes the format to be more readable.
+     *
+     * The initial format of LocalDateTime is YYYY-MM-DDTHH:MM:SS.
```
I like how clear your JavaDocs comments are.

hyc17003 commented on src/main/java/duke/Tag.java:
```diff
@@ -0,0 +1,5 @@
+package duke;
+
+public enum Tag {
```
I like the use of the tag, seems to simplify the code!

hyc17003 commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,58 @@
+     * Marks the task as complete.
+     * Toggles the isCompleted field to true.
+     */
+    public void markAsComplete() {
```
Following the Java coding standard, maybe this can be renamed to setCompleted?
```suggestion
    public void setCompleted() {
```

hyc17003 reviewed (commented):
LGTM! Maybe some variables/methods could be renamed to adhere to coding standards, but overall good job!

Gabriel4357 committed:
Add assertions to the code.

Hard to detect errors made by programmer.

Having assertions help detect bugs in code and test
current assumptions.

Let's,
*add assertions to check if our assumptions on
length of splitText is true
*add assertions to check if certain variables are
non null

Gabriel4357 committed:
Merge pull request #1 from Gabriel4357/branch-A-Assertions

Add assertions to the code.

Gabriel4357 committed:
Merge branch 'master' of https://github.com/Gabriel4357/ip

* 'master' of https://github.com/Gabriel4357/ip:
  Add assertions to the code.

Gabriel4357 committed:
Extract methods out partially

Gabriel4357 committed:
Extract out Commands for parse method

Parser class contains parse method which is extremely long
of 200 lines.

To ensure code quality, methods should avoid
being too long and convoluted.

Let's
*extract out all of the code into separate commands into
different classes

Gabriel4357 committed:
Refactor the code

Code quality is poor in certain parts of the code.

Makes it hard for readers to comprehend the code.

Let's
*refactor using Guard Clauses
*refactor using Reverse Conditional
*refactor by giving variables to boolean expressions

Gabriel4357 committed:
Merge branch 'master' into branch-A-CodeQuality

Gabriel4357 committed:
Merge pull request #2 from Gabriel4357/branch-A-CodeQuality

Branch a code quality

Gabriel4357 committed:
Merge branch 'master' into branch-BCD-Extension

* master:
  Refactor the code
  Extract out Commands for parse method
  Extract methods out partially
  Add assertions to the code.

Gabriel4357 committed:
Sort tasks based on Tag and Dates

Tasks are sorted based on recency of adding to the TaskList.

Being able to sort task based on the dates will be much more useful
and less messy in general.

Let's
*add two comparators to help us compare events
*add a sorting command to carry out the sort

Gabriel4357 committed:
Merge pull request #3 from Gabriel4357/branch-BCD-Extension

Branch bcd extension

Gabriel4357 committed:
Merge branch 'master' of https://github.com/Gabriel4357/ip

* 'master' of https://github.com/Gabriel4357/ip:
  Sort tasks based on Tag and Dates

Gabriel4357 committed:
Merge branch 'branch-A-CodeQuality' of https://github.com/Gabriel4357/ip into branch-A-CodeQuality

* 'branch-A-CodeQuality' of https://github.com/Gabriel4357/ip:
  Add assertions to the code.

Gabriel4357 committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:

Gabriel4357 committed:
Resize window and add image to docs

Gabriel4357 committed:
Add User Guide

Gabriel4357 committed:
Merge pull request #5 from Gabriel4357/branch-A-UserGuide

Add User Guide

Gabriel4357 committed:
Merge branch 'master' of https://github.com/Gabriel4357/ip

* 'master' of https://github.com/Gabriel4357/ip:
  Add User Guide

Gabriel4357 committed:
Fix user guide

Gabriel4357 committed:
Build a new jar file

Gabriel4357 committed:
Fix jar file

Gabriel4357 committed:
Fix gradle build
